lib/puppet/type/keystore.rb

```diff
@@ -0,0 +1,55 @@
+Puppet::Type.newtype(:keystore) do
+  desc 'generates an empty pkcs12 keystore'
+
+  ensurable
+
+  newparam(:keystore, :namevar => true) do
+    desc "Path to the keystore"
+    isrequired
+  end
+
+  newparam(:password_file) do
+    desc "Path to file containing the keystore password"
+    isrequired
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the keystore. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the keystore. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the keystore. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  autorequire(:file) do
+    [self[:password_file]]
+  end
+
+  def generate
+    file_opts = {
+      ensure: (self[:ensure] == :absent) ? :absent : :file,
```

Comment:
I thought that this opens a possibility where someone declares `ensure => 'purged',` (for example) and this ultimately (and incorrectly) gets interpreted as `:file`.

But, trying to reproduce the potential problem I see that I am prevented from making that mistake: `Error: Parameter ensure failed on Keystore[/etc/candlepin/certs/keystore]: Invalid value "purged". Valid values are present, absent. (file: /etc/puppetlabs/code/environments/production/modules/certs/manifests/candlepin.pp, line: 129)`

I then found https://github.com/puppetlabs/puppet/blob/67ad21b64237eb4f808844bdaa3c2b0e4cdef95c/lib/puppet/type.rb#L197 and https://github.com/puppetlabs/puppet/blob/67ad21b64237eb4f808844bdaa3c2b0e4cdef95c/lib/puppet/property/ensure.rb#L13-L30 so I am fully satisfied and understand the logic here.